File: src/main/java/nl/waterjeloen/quoridor/Location.java
package nl.waterjeloen.quoridor;

import java.util.Objects;

public class Location {
    public final int row;
    public final int column;

    public Location(int row, int column) {
        this.row = row;
        this.column = column;
    }

    public Location go(Direction direction) {
        final int newRow = row + direction.rowDelta;
        final int newColumn = column + direction.columnDelta;
        return new Location(newRow, newColumn);
    }

    public Location wall(Direction side) {
        final int newRow = row + (side.rowDelta - 1) / 2;
        final int newColumn = column + (side.columnDelta - 1) / 2;
        return new Location(newRow, newColumn);
    }

    public Location left() {
        return left(1);
    }

    public Location left(int amount) {
        return new Location(row, column - amount);
    }

    public Location right() {
        return right(1);
    }

    public Location right(int amount) {
        return new Location(row, column + amount);
    }

    public Location up() {
        return up(1);
    }

    public Location up(int amount) {
        return new Location(row - amount, column);
    }

    public Location down() {
        return down(1);
    }

    public Location down(int amount) {
        return new Location(row + amount, column);
    }

    public boolean isValid(int size) {
        return isValid(size, size);
    }

    public boolean isValid(int rowCount, int columnCount) {
        return isRowValid(rowCount) && isColumnValid(columnCount);
    }

    public boolean isRowValid(int rowCount) {
        return (row >= 0) && (row < rowCount);
    }

    public boolean isColumnValid(int columnCount) {
        return (column >= 0) && (column < columnCount);
    }

    @Override
    public boolean equals(Object object) {
        if (this == object) return true;
        if (object == null || getClass() != object.getClass()) return false;
        final Location location = (Location) object;
        return row == location.row && column == location.column;
    }

    @Override
    public int hashCode() {
        return Objects.hash(row, column);
    }

    @Override
    public String toString() {
        return "(" + row + "," + column + ")";
    }
}

File: src/main/scala/nl/zwaaltjes/quoridor/Location.scala
package nl.zwaaltjes.quoridor

final case class Location(row: Int, column: Int) {
  import Location.*

  def go(direction: Direction): Location = {
    val newRow = row + direction.rowDelta
    val newColumn = column + direction.columnDelta
    Location(newRow, newColumn)
  }

  def wall(side: Direction): Location = {
    val newRow = row + (side.rowDelta - 1) / 2
    val newColumn = column + (side.columnDelta - 1) / 2
    Location(newRow, newColumn)
  }

  def left(implicit amount: Int = 1): Location =
    Location(row, column - amount)

  def right(implicit amount: Int = 1): Location =
    Location(row, column + amount)

  def up(implicit amount: Int = 1): Location =
    Location(row - amount, column)

  def down(implicit amount: Int = 1): Location =
    Location(row + amount, column)

  def isValid(size: Int): Boolean =
    isValid(size, size)

  def isValid(rowCount: Int, columnCount: Int): Boolean =
    isRowValid(rowCount) && isColumnValid(columnCount)

  def isRowValid(rowCount: Int): Boolean =
    (row >= 0) && (row < rowCount)

  def isColumnValid(columnCount: Int): Boolean =
    (column >= 0) && (column < columnCount)

  override def toString: String =
    s"${columnString(column)}${rowString(row)}"
}

object Location {
  def rowString(row: Int): String =
    (row + 1).toString

  def columnString(column: Int): String =
    ('A' + column).toChar.toString
}